src/libcore/str/mod.rs

```diff
@@ -1074,26 +1075,82 @@ unsafe fn cmp_slice(a: &str, b: &str, len: usize) -> i32 {
 }
 
 /*
-Section: Misc
+Section: UTF-8 validation
 */
 
+// use truncation to fit u64 into usize
+const NONASCII_MASK: usize = 0x80808080_80808080u64 as usize;
+
+/// Return `true` if any byte in the word `x` is nonascii (>= 128).
+#[inline]
+fn contains_nonascii(x: usize) -> bool {
+    (x & NONASCII_MASK) != 0
+}
+
+/// Unchecked byte slice
+unsafe fn slice_unchecked(v: &[u8], from: usize) -> &[u8] {
+    debug_assert!(from <= v.len());
+    slice::from_raw_parts(v.as_ptr().offset(from as isize),
+                          v.len() - from)
+}
+
+/// Return the first index where a nonascii byte is located,
+/// or `None` if no was found.
+fn find_nonascii(v: &[u8]) -> Option<usize> {
+    let len = v.len();
+    let ptr = v.as_ptr();
+
+    let mut offset = 0;
+    if len >= 2 * usize::BYTES {
+        // search up to an aligned boundary
+        let align = (ptr as usize) & (usize::BYTES - 1);
+        if align > 0 {
+            offset = cmp::min(usize::BYTES - align, len);
+            if let Some(index) = v[..offset].iter().position(|elt| *elt >= 128) {
+                return Some(index);
+            }
+        }
+
+        while offset <= len - 2 * usize::BYTES {
+            unsafe {
+                let u = *(ptr.offset(offset as isize) as *const usize);
+                let v = *(ptr.offset((offset + usize::BYTES) as isize) as *const usize);
+
+                // break if there is a matching byte
+                let zu = contains_nonascii(u);
+                let zv = contains_nonascii(v);
+                if zu || zv {
+                    break;
+                }
+            }
+            offset += usize::BYTES * 2;
+        }
+    }
+
+    // find the byte after the point the loop stopped
```

Comment:
jawik10 is the first 10MB from the unzip of http://dumps.wikimedia.org/archive/2006/2006-07/jawiki/20061016/jawiki-20061016-pages-articles.xml.bz2